core/src/main/java/org/infinispan/container/impl/EntryFactoryImpl.java

```diff
@@ -84,7 +85,7 @@ public void init() {
             }
          } else if (isOwner || readEntry.isL1Entry()) {
             if (readEntry.canExpire()) {
-               CompletionStage<Boolean> expiredStage = expirationManager.handlePossibleExpiration(readEntry, segment, false);
+               CompletionStage<Boolean> expiredStage = expirationManager.handlePossibleExpiration(readEntry, segment, hasLock);
```

Comment:
It doesn't skip the max-idle expiration check. What it does is change the flags on the cache that the remove (max idle | lifespan) expired command is invoked on. For pessimistic transactions if it already has the lock we can't fire the expiration command and get a lock as it is already held.

https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/expiration/impl/TxClusterExpirationManager.java#L37

Reply:
I'm looking at https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/expiration/impl/ExpirationManagerImpl.java#L261
in pessimistic cache, `checkExpiredMaxIdle(ice, segment)` isn't invoked because `isWrite==true`

Reply:
That is because the entry wasn't expired. In this case we don't want to touch the entry on a write as it will be overwritten anyways.

Reply:
yeah, but a get+force_write_lock doesn't mean the key will be written in the transaction. Not sure if I'm making my concern clear...

Reply:
I made the following test: https://github.com/pruivo/infinispan/blob/8d5d98492df393c344fe2707dcd12dcf7498464f/core/src/test/java/org/infinispan/expiration/impl/MaxIdlePessimisticTxTest.java

* tx read + rollback, touches the key (it makes senses IMO)
* tx write + rollback, doesn't touch the key (it should IMO)
* tx read on expired key +  rollback, StackOverflowException (bug!) `org.infinispan.expiration.impl.ClusterExpirationManager.entryExpiredInMemory`
* tx write on expired key +  rollback, it is fine!

Reply:
So I am changing my mind on #2. Originally max idle didn't block the user read operation, but now it does so I don't believe it makes any sense any longer.

#3 is because we need to pass the hasLock variable on a read when using Pessimistic tx with WRITE_LOCK flag present. So it isn't just any tx read :)